src/test/scala/chiselTests/QueueFlushSpec.scala

```diff
@@ -0,0 +1,221 @@
+package chiselTests
+
+import org.scalacheck._
+
+import chisel3._
+import chisel3.testers.{BasicTester, TesterDriver}
+import chisel3.util._
+import chisel3.util.random.LFSR
+
+/** Test elements can be enqueued and dequeued
+ * 
+  * @param elements The sequence of elements used in the queue
+  * @param queueDepth The max number of entries in the queue
+  * @param bitWidth Integer size of the data type used in the queue
+  * @param tap Integer tap('seed') for the LFSR
+  * @param useSyncReadMem True uses SyncReadMem instead of Mem as an internal memory element
+  */
+class ThingsPassThroughFlushQueueTester(elements: Seq[Int], queueDepth: Int, bitWidth: Int, tap: Int, useSyncReadMem: Boolean) extends BasicTester {
+  
+  val q = Module(new Queue(UInt(bitWidth.W), queueDepth, hasFlush = true))  
+  q.io.flush.get := false.B
+  val elems = VecInit(elements.map(_.U))
+
+  val inCnt = Counter(elements.length + 1)
+  val outCnt = Counter(elements.length + 1)
+
+  q.io.enq.valid := (inCnt.value < elements.length.U)
+  q.io.deq.ready := LFSR(16)(tap)
+
+  q.io.enq.bits := elems(inCnt.value)
+  when(q.io.enq.fire()) {
+    inCnt.inc()
+  }
+  when(q.io.deq.fire()) {
+    //ensure that what comes out is what comes in
+    assert(elems(outCnt.value) === q.io.deq.bits)
+    outCnt.inc()
+  }
+  when(outCnt.value === elements.length.U) {
+    stop()
+  }
+}
+
+/** Test queue can flush at random times
+ * 
+  * @param elements The sequence of elements used in the queue
+  * @param queueDepth The max number of entries in the queue
+  * @param bitWidth Integer size of the data type used in the queue
+  * @param tap Integer tap('seed') for the LFSR
+  * @param useSyncReadMem True uses SyncReadMem instead of Mem as an internal memory element
+  */
+class QueueGetsFlushedTester (elements: Seq[Int], queueDepth: Int, bitWidth: Int, tap: Int, useSyncReadMem: Boolean) extends BasicTester {
+  val q = Module(new Queue(UInt(bitWidth.W), queueDepth, hasFlush = true))
+  val elems = VecInit(elements.map(_.U))
+
+  val inCnt = Counter(elements.length + 1)
+  val outCnt = Counter(elements.length + 1)
+  
+  val halfCnt = (queueDepth + 1)/2
+  val cycleCounter = Counter(elements.length + 1)
+  //testing a flush when 
+  val flush = LFSR(16)((tap + 3) % 16)
+  q.io.flush.get := flush
+  val flushRegister = RegInit(false.B)
+  flushRegister := flush
+  q.io.enq.valid := (inCnt.value < elements.length.U)
+  q.io.deq.ready := LFSR(16)(tap)
+  cycleCounter.inc() //counts every cycle
+
+  q.io.enq.bits := elems(inCnt.value)
+  when(q.io.enq.fire()) {
+    inCnt.inc()
+    inCnt.inc()
```

Comment:
what does it mean to have this twice here? Do you really mean to increment the counter twice?